common-npm-packages/kubernetes-common/kubelogin.ts

```diff
@@ -57,30 +58,42 @@ export class Kubelogin {
       await kubectaskLibTool.exec();
     } else if (authScheme.toLowerCase() == 'serviceprincipal') {
       const authType: string = taskLib.getEndpointAuthorizationParameter(connectedService, 'authenticationType', true);
-      let servicePrincipalKey: string = null;
       const servicePrincipalId: string = taskLib.getEndpointAuthorizationParameter(connectedService, 'serviceprincipalid', false);
       const tenantId: string = taskLib.getEndpointAuthorizationParameter(connectedService, 'tenantid', false);
 
       if (authType == 'spnCertificate') {
         taskLib.debug('certificate based endpoint');
         let certificateContent: string = taskLib.getEndpointAuthorizationParameter(connectedService, 'servicePrincipalCertificate', false);
-        servicePrincipalKey = path.join(taskLib.getVariable('Agent.TempDirectory') || taskLib.getVariable('system.DefaultWorkingDirectory'), 'spnCert.pem');
-        fs.writeFileSync(servicePrincipalKey, certificateContent);
+        let certificatePath: string = path.join(taskLib.getVariable('Agent.TempDirectory') || taskLib.getVariable('system.DefaultWorkingDirectory'), 'spnCert.pfx');
+        
+        // kubelogin works only with pfx. Hence converting pem to pfx
+        fs.writeFileSync(certificatePath, Kubelogin.convertToPFX(certificateContent), 'binary');
+
+        const kubectaskLibTool: toolRunner.ToolRunner = taskLib.tool(this.toolPath);
+        kubectaskLibTool.arg('convert-kubeconfig');
+        kubectaskLibTool.arg(['-l', 'spn']);
+        if (taskLib.getVariable('System.Debug')) {
+          kubectaskLibTool.arg(['--v', '20']);
+        }
+        await kubectaskLibTool.exec();
+
+        process.env['AAD_SERVICE_PRINCIPAL_CLIENT_ID'] = servicePrincipalId;

```

Comment:
Please consider `tasklib.setTaskVariable` option.

https://github.com/microsoft/azure-pipelines-task-lib/blob/master/node/docs/azure-pipelines-task-lib.md#tasksetTaskVariable

Reply:
I changed it to taskLib.SetVariable because taskLib.SetTaskVariable uses local vault that cannot be reached by kubelogin.